examples/open-web/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_bindgen::collections::{Map, Vector};
+use near_bindgen::{env, ext_contract, near_bindgen, Promise, PromiseOrValue};
+use serde::{Deserialize, Serialize};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+type PublicKey = Vec<u8>;
+type AppId = String;
+type Key = String;
+type Value = String;
+type Message = String;
+type AccountId = String;
+
+const GAS: u64 = 1_000_000_000_000_000;
+
+const APP_METHODS: &[u8] = b"set,remove,pull_message,send_message";
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct UserData {
+    /// App specific PublicKey -> App ID prefix.
+    pub app_keys: Map<PublicKey, AppId>,
+
+    /// Number of app keys per app_id.
+    pub num_app_keys: Map<AppId, u64>,
+
+    /// Received messages. Should be a queue, but whatever.
+    pub messages: Map<AppId, Vector<WrappedMessage>>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+pub struct WrappedMessage {
+    pub sender: AccountId,
+    pub message: Message,
+    pub time: u64,
+}
+
+fn verify_app_id(app_id: &AppId) {
+    if app_id.len() < 2 || app_id.len() > 64 {
+        env::panic(b"App ID length should be between 2 and 64 characters");
+    }
+    for c in app_id.bytes() {
+        match c {
+            b'a'..=b'z' => (),
+            b'0'..=b'9' => (),
+            b'-' | b'_' | b'.' => (),
+            _ => env::panic(
+                b"Unsupported character in the app ID. Only allowed to use `-.|` and 0-9 a-z",
+            ),
+        }
+    }
+}
+
+fn app_key(app_id: &AppId, key: &Key) -> Vec<u8> {
+    let mut res = Vec::with_capacity(app_id.len() + key.len() + 1);
+    res.extend_from_slice(app_id.as_bytes());
+    res.push(b':');
+    res.extend_from_slice(key.as_bytes());
+    res
+}
+
+// One can provide a name, e.g. `ext` to use for generated methods.
+#[ext_contract(remote)]
+pub trait RemoteSelf {
+    fn post_message(&mut self, app_id: AppId, message: Message);
+}
+
+#[near_bindgen(init => new)]
```

Comment:
What does this mean?

Reply:
It's a macro with param. Indicates that we use `new` as a constructor for the object. `near-bindgen` behind the scene always deserializes the state of the object (from persistent storage) before calling a public function and serializes the state back at the end of the function execution.